File: day04/main.py
filename = "./input"

with open(filename, "r") as file:
    lines = file.readlines()
lines = [line.strip() for line in lines]


containments = 0
overlaps = 0
for line in lines:
    print("")
    elf1, elf2 = line.split(",")
    elf1 = [int(e) for e in elf1.split("-")]
    elf2 = [int(e) for e in elf2.split("-")]

    if (elf1[0] >= elf2[0] and elf1[1] <= elf2[1]) or (
        elf2[0] >= elf1[0] and elf2[1] <= elf1[1]
    ):
        containments += 1

    # visualization
    print(
        "".join(
            ["." for _ in range(0, elf1[0])]
            + ["#" for _ in range(elf1[0], elf1[1] + 1)]
            + ["." for _ in range(elf1[1], 101)]
        )
    )
    print(
        "".join(
            ["." for _ in range(0, elf2[0])]
            + ["#" for _ in range(elf2[0], elf2[1] + 1)]
            + ["." for _ in range(elf2[1], 101)]
        )
    )

    if elf1[1] >= elf2[0] and elf1[0] <= elf2[1]:
        overlaps += 1
        print(elf1, elf2, "OVERLAP!")
    else:
        print(elf1, elf2)


print(f"Puzzle 1: {containments}")
print(f"Puzzle 2: {overlaps}")

File: day04/main.go
package main

import (
	aoc "aoc/utils"
	"fmt"
	"strconv"
	"strings"
)

const filepath = "./input"

func parseLine(line string) [4]int {
	var intervals [4]int
	for i, section := range strings.Split(line, ",") {
		for j, intervalStr := range strings.Split(section, "-") {
			interval, err := strconv.Atoi(intervalStr)
			aoc.Handle(err)
			intervals[(i*2)+j] = interval
		}
	}
	return intervals
}

func main() {
	lines, err := aoc.ReadLines(filepath)
	aoc.Handle(err)

	var containments int
	var overlaps int
	for _, line := range lines {
		e := parseLine(line)

		if (e[0] >= e[2] && e[1] <= e[3]) || (e[2] >= e[0] && e[3] <= e[1]) {
			containments++
		}

		if e[1] >= e[2] && e[0] <= e[3] {
			overlaps++
		}
	}

	fmt.Println("Puzzle 1:", containments)
	fmt.Println("Puzzle 2:", overlaps)
}
